### PR TITLE
[RFR] utils.browser: POC to extract and robust-ify wharf heartbeats/renewal

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
this pr externalizes wharf heartbeats and tries to make delay control, exiting and restarting more safe
all renew calls will in all cases happen in the managing thread

delay control and communication is implemented by threading.Event objects

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 8d82ea53d913d50113f81672e2a5e5a244d8ac05:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

File: utils/browser.py [hunk 1]
```diff
@@ -47,11 +47,43 @@ def _load_firefox_profile():
...
+        self._trigger.set()
+        self._done.wait()
+
+
 class Wharf(object):
     def __init__(self, wharf_url):
         self.wharf_url = wharf_url
         self.docker_id = None
-        self.renew_timer = None
+        self.heartbear = None
```
Comment (nellyc) [inline utils/browser.py]: heartbeat

**Comment by jaryn** [comment]
Umn, this is way more complex then the Timer version. What exactly is this solving? Maybe the timer has a problem when the time interval is too small?

**Comment by dajoRH** [comment]
Lint report for commit a49055631f22886f716d4678da866bf9b8b7d7e2:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by RonnyPfannschmidt** [comment]
@jaryn this code removes possible issues created by renew interacting with the timer by ensuring there is one thread responsible for renewal, and only it runs the code

that way a call to renew, just will short circut the wait for the next renew in the thread

**Comment by jaryn** [comment]
```
.cfme_tests/lib/python2.7/site-packages/navmazing/__init__.py:104: in navigate
    return nav(cls_or_obj, self).go(0, *args, **kwargs)
utils/appliance/implementations/ui.py:347: in go
    self.pre_navigate(_tries, *args, **kwargs)
utils/appliance/implementations/ui.py:151: in pre_navigate
    self.appliance.browser.open_browser()
utils/appliance/implementations/ui.py:396: in open_browser
    return manager.ensure_open(url_key=None)
utils/browser.py:276: in ensure_open
    self.factory.renew()
utils/browser.py:231: in renew
    self.wharf.renew()
utils/browser.py:121: in renew
    self.heartbeat.manual_beat()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = HeartBeat(_thread_name=None)

    def manual_beat(self):
>       self._trigger.set()
E       AttributeError: '_Condition' object has no attribute 'set'
```

**Comment by RonnyPfannschmidt** [comment]
closing as failed experiment

**Pull request closed without merging**